README.md
# Docker-Rails
[![Circle CI](https://circleci.com/gh/ComboStrikeHQ/docker-rails.svg?style=svg)]
(https://circleci.com/gh/ComboStrikeHQ/docker-rails)

An opinionated docker image for running Rails apps in production.

Uses Puma, Sidekiq, Clockwork and rails_migrate_mutex.
Everything is optional.

## Quick Example
- Create a new Rails app (`$ rails new my_blog`)
- Create a `Dockerfile` in the app root with this content: `FROM combostrikehq/docker-rails:latest`
- Add `gem 'puma'` to the `Gemfile` and run `bundle install`
- Build the docker container: `$ docker build -t my_blog .`
- Run the container: `$ docker run -p 8080:8080 -e RAILS_ENV=development my_blog`
- Open `http://localhost:8080`
- Enjoy!

This example is for demo purposes only.
Docker-Rails is intended for production use, not local development.
Please read the whole README before using.

## Docker Image
The image is built by us, squashed into a single layer and pushed to
[Docker Hub](https://registry.hub.docker.com/u/combostrikehq/docker-rails/).

To use it in your app, create a `Dockerfile` with the following content:

```docker
FROM combostrikehq/docker-rails:<VERSION>
```

and replace `<VERSION>` with the [current version number](https://hub.docker.com/r/combostrikehq/docker-rails/tags/).
You can also use the `latest` version tag which would give you the latest unreleased version. This
might be useful if you want to test against the bleeding edge in your testing/beta environment.

From version 3.0 docker-rails uses ruby 2.5. If you still use ruby 2.4 you have to use docker-rails
versions 2.x. We also provide `ruby-2.5` and `ruby-2.4` tags that will point to the latest released
docker-rails version for that ruby version.

Add more docker steps if needed (usually not), build and deploy to your infrastructure.

We use [eb_deployer](https://github.com/ThoughtWorksStudios/eb_deployer) to deploy our apps
to [AWS Elastic Beanstalk](https://aws.amazon.com/de/elasticbeanstalk/).

## Contents

### Rails App Server / Puma
[Puma](http://puma.io/) is used as the web server for your app.
For this to work properly, please add the following gems to your Gemfile:

```ruby
gem 'puma'
gem 'rack-timeout', group: :production
gem 'rails_12factor', group: :production  # only for Rails < 5
```

You can configure Puma:

Environment Variable | Description | Default Value
--- | --- | ---
PUMA_WORKERS | Number of worker processes | [number of CPU cores]
PUMA_MAX_THREADS | Maximum number of threads per process | 16
PUMA_MIN_THREADS | Minimum number of threads per process | PUMA_MAX_THREADS

You can also create a `config/puma.rb` file in your app that contains additional configuration.

Have a look at the [Puma configuration file](base/puma.rb) for more details.

### Sidekiq
[Sidekiq](http://sidekiq.org/) is used for processing background jobs. A sidekiq process
will be automatically started when the `sidekiq` gem is present in your app. Using sidekiq is
optional.

You can configure sidekiq:

Environment Variable | Description | Default Value
--- | --- | ---
SIDEKIQ_THREADS | Number of threads | 16

You can also create a `config/sidekiq.yml` file in your app that contains additional configuration.

The default configuration enables the `default` and `mailers` queues.

### Clockwork
[Clockwork](https://github.com/tomykaira/clockwork) is used for scheduling background jobs.
A clockwork process will be automatically started when the `clockwork` gem is present in your app.
Using clockwork is optional.

You need to create a `config/clockwork.rb` file using the following schema:
```ruby
module Clockwork
  every(1.hour, 'my_job') { MyWorker.perform_async }
end
```

Note that the clockwork process is started in every container. We recommend using
[sidekiq-unique-jobs](https://github.com/mhenrixon/sidekiq-unique-jobs) to ensure that jobs
are scheduled only once.

### Custom services

You can run additional services by putting their executable files
in the application `bin/services` folder.

### Rails Migrations / rails_migrate_mutex
[rails_migrate_mutex](https://github.com/combostrikehq/rails_migrate_mutex) is used to run Rails
migrations. If you have the gem installed in you app, migrations are automatically run on
container startup. Using rails_migrate_mutex is optional.

### Ruby / Packages
The latest ruby version is included in the container. It is compiled from source and uses
[jemalloc](http://www.canonware.com/jemalloc/). The latest rubygems and bundler versions
are installed as well.

The following tools are installed: `git`, `nodejs`, `imagemagick`, `curl`, `bower`.
Development packages for compiling the following gems are included:
`sqlite3`, `pg`, `mysql2`, `nokogiri`.

Have a look at the [Base Dockerfile](base/Dockerfile) for more details.

### App Installation / Docker ONBUILD
The final docker image has ONBUILD hooks that set up your application.
It installs all gems except those in the `development` and `test` groups.

If the `sprockets`
gem is present (default for Rails apps), it also precompiles assets. If you need any environment
variables set for compiling assets, you can specify them in a `.env` file in your app root. This
file will be sourced before running the task. Newer versions of sprockets will also automatically
created gzipped versions of the assets. Rails then uses these static gzip files to serve assets.

Have a look at the [ONBUILD Dockerfile](onbuild/Dockerfile) for more details.

### Environment Variables
All environment variables are passed on the application. Both `RAILS_ENV` and `RACK_ENV` are
set to `production` and should not be changed.

Both `RAILS_SERVE_STATIC_FILES` and
`RAILS_LOG_TO_STDOUT` are set, so you
[don't need to](https://github.com/heroku/rails_12factor#rails-5-and-beyond)
include the `rails_12factor` gem anymore if you're using Rails 5 or above and have updated
your `config/environments/production.rb` file.

## License

MIT, see LICENSE.txt

## Contributing

Please create a new GitHub Issue if you encounter a bug or have a feature request.

Feel free to fork and submit pull requests!


CHANGELOG.md
# Changelog

## 2.8.5 (2021-12-09)

- Explicitly set sidekiq environment to `RAILS_ENV`

## 2.8.4 (2019-07-04)

- Implement workaround for `HOME` being ignored in `phusion/baseimage`

## 2.8.3 (2019-05-24)

- Port to CircleCI 2.0
- Make it work with `bundle package` (run before building the container), so
  fetching gems from within the container can be avoided. This is useful if you
  want to use dependencies from private repositories and don't want to give
  your container itself access to them.

## 2.8.2 (2019-02-05)

- Pass along `BUGSNAG_API_KEY` and `BUGSNAG_APP_VERSION` to asset precompiling
  step
- Update rubygems (3.0.2)
- Update bundler (1.8.2)

## 2.8.1 (2018-11-21)

- Upgrade Ruby (2.5.3)
- Upgrade rubygems (2.7.8)
- Upgrade bundler (1.17.1)

## 2.7.3 (2018-02-15)

- Add `file` command

- Upgrade Ruby (2.5.0)
- Upgrade rubygems (2.7.4)
- Upgrade bundler (1.16.1)
- Upgrade NodeJS (8.9.4 LTS)
- Upgrade phusion/baseimage (0.10.0)

## 2.7.2 (2018-01-25)

- Introduce ruby-2.4 version tag

## 2.7.1 (2018-01-02)

- Upgrade Ruby (2.4.3)
- Upgrade rubygems (2.6.14)
- Upgrade bundler (1.16.1)

## 2.7.0 (2017-12-22)

- Allow to run additional application services

## 2.6.2 (2017-09-13)

- Add `LIBRATO_AUTORUN` to sidekiq environment

## 2.6.1 (2017-07-20)

- Close Active Record connections [PR #15](https://github.com/ComboStrikeHQ/docker-rails/pull/15)

## 2.6.0 (2017-01-26)

- Upgrade rubygems
- Upgrade bundler
- Upgrade bower
- Switch to NodeJS LTS
- Add support for vendored gems

## 2.5.0 (2016-09-22)

- Add support for Rails 5 native 12-factor config
- Remove puma request logging
- Wait for syslog-ng to start

## 2.4.2 (2016-09-19)

- Fix application server logs missing

## 2.4.1 (2016-09-01)

- Fix asset precompilation when pg gem is not present

## 2.4.0 (2016-08-31)

- Add support for mysql2 gem
- Clean up README

## 2.3.0 (2016-07-26)

- Upgrade baseimage, rubygems, bundler and node

## 2.2.1 (2016-06-03)

- Add proper exit handling for Clockwork

## 2.2.0 (2016-05-18)

- Reduce number of layers in ONBUILD Dockerfile
- Upgrade ruby, rubygems, bundler and nodejs
- Fix graceful shutdown for Puma workers

## 2.1.1 (2016-05-13)

- Allow writing to app directory during ONBUILD

## 2.1.0 (2016-04-06)

- Upgrade rubygems, nodejs and bower
- Do not fail in puma boot when there is no ActiveRecord

## 2.0.1 (2016-02-08)

- Fix bug where everything is logged twice

## 2.0.0 (2016-02-08) BREAKING CHANGES, see README

- Remove logentries and NewRelic server monitor

## 1.6.0 (2016-02-02)

- Upgrade bower to 1.7.6
- Upgrade rubygems to 2.5.2
- Upgrade nodejs to 5.5.0

## 1.5.0 (2016-01-15)

- Upgrade to Ruby 2.3.0
- Upgrade to nodejs 5.3.1 and bower 1.7.2
- Ruby: Enable shared lib and remove static lib
- Speed up tests by removing one `bundle install` call

## 1.4.0 (2015-12-23)

- Allow writing to vendor directory during ONBUILD
- Set HOME in ONBUILD Dockerfile
- Add bower
- Use recent nodejs version
- Update to baseimage-docker 0.9.18
- Remove manual gzipping of assets - newer sprockets versions do that automatically
- Lock rubygems and bundler version

## 1.3.0 (2015-12-17)

- Upgrade to Ruby 2.2.4

## 1.2.2 (2015-11-27)

- Make some ONBUILD commands regular ones

## 1.2.1 (2015-11-12)

- Add prefix to docker tags (for development only)
- Use COPY in ONBUILD Dockerfile

## 1.2.0 (2015-10-28)

- Add gzip compression for asset files

## 1.1.1 (2015-10-10)

- Also log stderr output for all processes

## 1.1.0 (2015-10-09)

- Add clockwork support

## 1.0.2 (2015-10-09)

- Fix sidekiq default queues precedence

## 1.0.1 (2015-10-08)

- Fix custom `puma.rb` configuration

## 1.0.0 (2015-10-08)

- Initial release


base/puma.rb
# frozen_string_literal: true

workers ENV.fetch('PUMA_WORKERS', `nproc`).to_i

max_threads = ENV.fetch('PUMA_MAX_THREADS', 16).to_i
min_threads = ENV.fetch('PUMA_MIN_THREADS', max_threads).to_i
threads min_threads, max_threads

bind 'tcp://0.0.0.0:8080'
preload_app!

# Don't wait for workers to finish their work. We might have long-running HTTP requests.
# But docker gives us only 10 seconds to gracefully handle our shutdown process.
# This settings tries to shut down all threads after 2 seconds. Puma then gives each thread
# an additional 5 seconds to finish the work. This adds up to 7 seconds which is still below
# docker's maximum of 10 seconds.
# This setting only works on Puma >= 3.4.0.
force_shutdown_after 2 if respond_to?(:force_shutdown_after)

on_worker_boot do
  ActiveRecord::Base.establish_connection if defined?(ActiveRecord)
end

# As we are preloading our application and using ActiveRecord
# it's recommended that we close any connections to the database here to prevent connection leakage
# This rule also applies to any connections to external services (Redis, databases, memcache, ...)
# that might be started automatically by the framework.
before_fork do
  ActiveRecord::Base.connection_pool.disconnect! if defined?(ActiveRecord)
end

custom_config = '/home/app/webapp/config/puma.rb'
instance_eval(File.read(custom_config)) if File.exist?(custom_config)
